### PR TITLE
Migrate Image::_data to std::unique_ptr

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
Related to #3146

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
If we had the opportunity to migrate to C++14 at least, then we will be able to use std::make_unique (for convenience) and std::exchange (to simplify move constructors and get rid of inherited swap() methods, std::swap doesn't work with literals for obvious reasons, so now we need to initialize members first and the perform swap).

**Comment by oleg-derevenetz** [comment]
If we could use C++14 features here, then we would use std::exchange directly in member initializer lists of move constructors for members of primitive types instead of this funny combination of member initializer list + std::swap then.

File: src/engine/image.cpp [hunk 1]
```diff
@@ -402,87 +401,78 @@ namespace fheroes2
...
     Image::Image( Image && image_ )
         : _width( 0 )
         , _height( 0 )
-        , _data( nullptr )
+        , _data( std::move( image_._data ) )
         , _singleLayer( false )
     {
-        swap( image_ );
-    }
+        using namespace std;
```
Comment (ihhub) [inline src/engine/image.cpp:415]: Please always avoid `using namespace`.

File: src/engine/image.cpp [hunk 2]
```diff
@@ -402,87 +401,78 @@ namespace fheroes2
...
         }
 
         return *this;
     }
 
     Image & Image::operator=( Image && image_ )
     {
-        swap( image_ );
+        if ( this != &image_ ) {
+            using namespace std;
```
Comment (ihhub) [inline src/engine/image.cpp:434]: Please always avoid `using namespace`.

File: src/engine/image.cpp [hunk 3]
```diff
@@ -402,87 +401,78 @@ namespace fheroes2
...
-        return _data;
+        return _data.get();
     }
 
     uint8_t * Image::transform()
     {
-        return _data + _width * _height;
+        if ( empty() ) {
+            return nullptr;
+        }
```
Comment (ihhub) [inline src/engine/image.cpp:459]: Please remove this check. It's caller responsibility to check if the image is not empty.

File: src/engine/image.cpp [hunk 4]
```diff
@@ -402,87 +401,78 @@ namespace fheroes2
...
+
+        return _data.get() + _width * _height;
     }
 
     const uint8_t * Image::transform() const
     {
-        return _data + _width * _height;
+        if ( empty() ) {
+            return nullptr;
+        }
```
Comment (ihhub) [inline src/engine/image.cpp:468]: Please remove this check. It's caller responsibility to check if the image is not empty.

File: src/engine/image.cpp [hunk 5]
```diff
@@ -491,46 +481,67 @@ namespace fheroes2
...
-    void Image::swap( Image & image )
+    void Image::copy( const Image & image )
     {
-        // We shouldn't swap different types of images.
+        // We shouldn't copy different types of images.
         assert( _singleLayer == image._singleLayer );
 
-        std::swap( _width, image._width );
-        std::swap( _height, image._height );
+        if ( !image._data || image._width <= 0 || image._height <= 0 ) {
```
Comment (ihhub) [inline src/engine/image.cpp:529]: The first check is enough. width and height can't be less or equal 0 if data is present.

File: src/engine/image.cpp [hunk 6]
```diff
@@ -545,31 +556,46 @@ namespace fheroes2
...
     }
 
-    Sprite & Sprite::operator=( Sprite && image )
+    Sprite & Sprite::operator=( Sprite && sprite )
     {
-        swap( image );
+        if ( this != &sprite ) {
+            Image::operator=( std::move( sprite ) );
+
+            using namespace std;
```
Comment (ihhub) [inline src/engine/image.cpp:593]: Please always avoid `using namespace`.

File: src/engine/image.cpp [hunk 7]
```diff
@@ -545,31 +556,46 @@ namespace fheroes2
...
+
         return *this;
     }
 
-    Sprite & Sprite::operator=( Sprite && image )
+    Sprite & Sprite::operator=( Sprite && sprite )
     {
-        swap( image );
+        if ( this != &sprite ) {
+            Image::operator=( std::move( sprite ) );
```
Comment (ihhub) [inline src/engine/image.cpp:591]: After this line `x` and `y` will be always 0.

File: src/engine/image.cpp [hunk 8]
```diff
@@ -491,46 +481,67 @@ namespace fheroes2
...
     void Image::reset()
     {
         if ( !empty() ) {
-            const size_t totalSize = static_cast<size_t>( _width * _height );
-            std::fill( image(), image() + totalSize, 0 );
-            std::fill( transform(), transform() + totalSize, 1 ); // skip all data
+            const size_t size = static_cast<size_t>( _width * _height );
+
+            std::fill( image(), image() + size, 0 );
+            std::fill( transform(), transform() + size, 1 ); // skip all data
```
Comment (ihhub) [inline src/engine/image.cpp:520]: Can we please leave this code as it is as here's no changes at all?

File: src/engine/image.cpp [hunk 9]
```diff
@@ -491,46 +481,67 @@ namespace fheroes2
     void Image::fill( uint8_t value )
     {
         if ( !empty() ) {
-            const size_t totalSize = static_cast<size_t>( _width * _height );
-            std::fill( image(), image() + totalSize, value );
-            std::fill( transform(), transform() + totalSize, 0 );
+            const size_t size = static_cast<size_t>( _width * _height );
+
+            std::fill( image(), image() + size, value );
+            std::fill( transform(), transform() + size, 0 );
```
Comment (ihhub) [inline src/engine/image.cpp:487]: Can we please leave this code as it is as here's no changes at all?

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @oleg-derevenetz , I left multiple comments here. Could you please take a look?

**Comment by ihhub** [comment]
> 
> 
> If we could use C++14 features here, then we would use std::exchange directly in member initializer lists of move constructors for members of primitive types instead of this funny combination of member initializer list + std::swap then.
> 
> P.S. Also I don't see why we can't use move/swap operations in case of different image types (in regard to singleLayer property). It's an exchange after all, not a copy.

singleLayer images are special images, mostly made as a rendering related images such as Display. Such images should not be moved anywhere else. This is why we have this assertion.

**Comment by oleg-derevenetz** [comment]
Regarding using namespace - it's a standard practice with std::move to ensure that if there are some custom overloads of swap for any type, they are able to be found by ADL, and if not, std::swap will be used. It's explained more in detail here, for example:

https://stackoverflow.com/questions/14402990/should-you-overload-swap-in-the-std-namespace/14403772#14403772

But, if you insist, I can remove using namespace, although it will not be according to the standard :)

**Comment by ihhub** [comment]
> 
> 
> Hi @ihhub
> 
> Regarding `using namespace` - it's a standard practice with `std::swap` to ensure that if there are some custom overloads of `swap` for any type, they are able to be found by ADL, and if not, `std::swap` will be used. It's explained more in detail here, for example:
> 
> https://stackoverflow.com/questions/14402990/should-you-overload-swap-in-the-std-namespace/14403772#14403772
> 
> But, if you insist, I can remove `using namespace`, although it will not be quite up to the standard :)

Hi @oleg-derevenetz , we are calling std::swap for base type so it's totally fine to do so. Also in our code we do not use `using namespace` (except 2 places for AI and one old place) so let's please just use std::swap as it's an equivalent of `using namespace std` and then call `swap`.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub 

> After this line x and y will be always 0.

Why do you think so?

https://godbolt.org/z/PP7o4jMGr

Am I missing something?

**Comment by ihhub** [comment]
> 
> 
> Hi @ihhub
> 
> > After this line x and y will be always 0.
> 
> Why do you think so?
> 
> https://godbolt.org/z/PP7o4jMGr
> 
> Am I missing something?

You're right! My bad :(

**Comment by oleg-derevenetz** [comment]
Hi @ihhub

> singleLayer images are special images, mostly made as a rendering related images such as Display. Such images should not be moved anywhere else. This is why we have this assertion.

OK, then I re-add this assert in move constructor & move assignment.

**Comment by ihhub** [comment]
@oleg-derevenetz , I actually tested the whole game without zeroing for 3 hours and I didn't catch none. It was false alarm me being sleepy at midnight... so we're good to remove the line.

**Comment by oleg-derevenetz** [comment]
@ihhub 

>  I actually tested the whole game without zeroing for 3 hours and I didn't catch none. It was false alarm me being sleepy at midnight... so we're good to remove the line.

OK, I'll remove this now. BTW, I'm curious, how big is the gain of moving to moving :))) As I can see (old-school printfs) there are very few copy operations now, most operations are move operations.

**Comment by ihhub** [comment]
Good job, @oleg-derevenetz ! Thanks for improving the code!

**Pull request merged**